### ISSUE TITLE
Navigation bar back button interception is not implemented

### ISSUE THREAD

**Issue by escamoteur** [daniel-luberda/DLToolkit.PageFactory#6]
I tried to override PageFactoryPopping, but it's not called when I press the backbutton:

``` c#
namespace BREADy.Views
{
    public partial class IngredientPage : PFContentPage<IngredientViewModel>
    {

        public IngredientPage ()
        {
            InitializeComponent();
        }

        public override bool PageFactoryPopping()
        {
            ViewModel.OnPagePopping();
            return base.PageFactoryPopping();
        }
    }
}
```

Do I need to initialize something for this to work?
Ho can I handle the backbutton specifically not just popping of a page?

**Comment by daniel-luberda** [comment]
```
Back button != popping

popping =  page.PopPage(); 
```

I tried to add back navigation bar button handling, but it's not an easy task on current Xamarin.Forms version. Whats more, I found a strange navigation Xamarin.Forms navigation stack behavior. I tried to contact Xamarin, but didn't get any response. Read more here:

> While doing my small MVVM toolkit I noticed something strange in Forms Navigation. Push a page to NavigationStack, then. 
> Pop it (PopAsync method) = it isn't in NavigationStack anymore
> Tap navigation back button = it's still in NavigationStack, until another page is pushed
> What's the reason for this inconsistency? Second question. Why we don't have an event for navigating back through navigation page?

If you have any clues how could we implement that - contact me. It's on my high priority list, I definitely want to implement that but I'm still looking how to achieve it.

Some info: 
- https://forums.xamarin.com/discussion/46983/how-to-intercept-navigation-bar-back-button-clicked-in-xamarin-forms
- http://stackoverflow.com/questions/31696595/how-to-intercept-navigation-bar-back-button-clicked-in-xamarin-forms

**Comment by daniel-luberda** [comment]
I didn't check if anything has changed in this matter for a while. I'll try to make a little investigation and check if I could implement it this time.

**Comment by escamoteur** [comment]
HI Daniel,

just wanted to change my PFContentPage to a NavigationPage as I need an Actionbar on a ModalPage, which is not available on a ContentPage.

Unfortunately PFNavigationPage does not take a ViewModel as Template.

What should I do?

Best

Thomas

Von: Daniel Luberda [mailto:notifications@github.com] 
Gesendet: Dienstag, 12. Januar 2016 00:47
An: daniel-luberda/DLToolkit.PageFactory DLToolkit.PageFactory@noreply.github.com
Cc: escamoteur thomas@burkharts.net
Betreff: Re: [DLToolkit.PageFactory] Navigation interception does not work (#6)

I didn't check if anything has changed in this matter for a while. I'll try to make a little investigation and check if I could implement it this time.

**Comment by daniel-luberda** [comment]
It does. There are two variants: 

``` C#
public class PFNavigationPage : NavigationPage<PFNavigationPage.NavigationPageViewModel>
```

and 

``` C#
public abstract class NavigationPage<TViewModel> : Xamarin.Forms.NavigationPage, IBasePage<TViewModel> where TViewModel : class, INotifyPropertyChanged
```

**Comment by daniel-luberda** [comment]
Oh no... it's a typo bug :D missed `PF` part :P Will have to fix that and make the old one `Obsolete` for API consistency. I don't understand why I didn't notice it up to this day as I'm using it quite often! Thanks!

**Comment by daniel-luberda** [comment]
Fixed: https://github.com/daniel-luberda/DLToolkit.PageFactory/commit/4fa83a5af83f74fd626993c03ee3c3ca4ec3487f

**Comment by escamoteur** [comment]
Hi Daniel,

if I just change PFContentPage to PFNavigationPage not contents is renedered anymore. I guess that’s because NavigationPage is normaly only a container for Contentpages.

How do I add a contentpage tot he Navigationpage when using PageFactory? You know I only need the Navigationpage to have the Actionbar on a Modal Contentpage.

Btw. You wrote you are only a hobbyist. I can’t believe that. You must be a professional.

I you should add a donate button to your Github CachedImage site as there is this huge positive feedback on the forum to it.

All the best

Thomas

Von: Daniel Luberda [mailto:notifications@github.com] 
Gesendet: Dienstag, 12. Januar 2016 23:27
An: daniel-luberda/DLToolkit.PageFactory DLToolkit.PageFactory@noreply.github.com
Cc: escamoteur thomas@burkharts.net
Betreff: Re: [DLToolkit.PageFactory] Navigation bar back button interception is not implemented (#6)

Fixed:  https://github.com/daniel-luberda/DLToolkit.PageFactory/commit/4fa83a5af83f74fd626993c03ee3c3ca4ec3487f 4fa83a5

**Comment by daniel-luberda** [comment]
@escamoteur Thanks for such warm words :) Yes, I'm a professional, but love side-projects a lot.

> How do I add a contentpage tot he Navigationpage when using PageFactory? You know I only need the Navigationpage to have the Actionbar on a Modal Contentpage.

It's very similar to pure Xamarin.Forms (inside Navigation Page constructor): 

``` C#
// That way you can get instance of page by using PageFactory cache:
CurrentPage = PageFactory.GetPageFromCache<YourViewModel>();

// OR: (that way it won't be cached in PageFactory cache)
CurrentPage = PageFactory.GetPageAsNewInstance<YourViewModel>();
```

**Comment by daniel-luberda** [comment]
BTW: I am thinking about adding a class like this: 

``` C#
public abstract class PFNavigationPage<TNavigationViewModel, TPageViewModel> : Xamarin.Forms.NavigationPage, IBasePage<TViewModel> where TViewModel : class, INotifyPropertyChanged
```

which will automatically call: `CurrentPage = PageFactory.GetPageFromCache<TPageViewModel>();`

but I have to think if it wouldn't be too confusing.

**Comment by escamoteur** [comment]
@escamoteur https://github.com/escamoteur  Thanks for such warm words :)

How do I add a contentpage tot he Navigationpage when using PageFactory? You know I only need the Navigationpage to have the Actionbar on a Modal Contentpage.

It's very similar to pure Xamarin.Forms (inside Navigation Page constructor): 

CurrentPage = PageFactory.GetPageFromCache<YourViewModel>();
// OR: (that way it won't be cached)
CurrentPage = PageFactory.GetPageAsNewInstance<YourViewModel>();

Yes that is clear, this only gives me the Navigation Page without any content. But my intention is getting a NavigationPage with a Contentpage that is linked tot he viewmodel.

Otherwise I will have to register a ViewModel just for the NavigationPage and another one for the actual content and then Push first the NavigationPage and then Push the Contentpage.

**Comment by daniel-luberda** [comment]
> Otherwise I will have to register a ViewModel just for the NavigationPage and another one for the actual content and then Push first the NavigationPage and then Push the Contentpage.

That is how this should be done. It's because Navigation Page `CurrentPage` can be changed dynamically to different pages, so navigation page view model should be independent of other pages view models. 

But, you would not need to push anything twice. Just push Navigation Page and that's all. If you have ContentPage set as an `CurrentPage` of NavigationPage, it will  be shown automatically. It's a standard Xamarin.Forms behavior.

**Comment by escamoteur** [comment]
Not sure if I completely understand what would be the benefit?

Von: Daniel Luberda [mailto:notifications@github.com] 
Gesendet: Mittwoch, 13. Januar 2016 12:28
An: daniel-luberda/DLToolkit.PageFactory DLToolkit.PageFactory@noreply.github.com
Cc: escamoteur thomas@burkharts.net
Betreff: Re: [DLToolkit.PageFactory] Navigation bar back button interception is not implemented (#6)

BTW: I am thinking about adding a class like this: 

public abstract class PFNavigationPage<TNavigationViewModel, TPageViewModel> : Xamarin.Forms.NavigationPage, IBasePage<TViewModel> where TViewModel : class, INotifyPropertyChanged

which will automatically call: CurrentPage = PageFactory.GetPageFromCache< TPageViewModel >();

**Comment by daniel-luberda** [comment]
@escamoteur Not much, but it would handle your case. CurrentPage view model would be also a NavigationPage view model. And by looking at Navigation Page definition, it would be obvious that for this navigation page, current page wouldn't be changed at runtime.

**Comment by escamoteur** [comment]
Right, especially as you can’t set CurrentPage directly. I have to Push the contentpage tot he NavigationPage

This ist he way I git it to work, but I guess this might not be the best solution

```
    public IngredientsMainViewModel()

    {

        var navPage = (BREADy.Views.NavigationPage) PageFactory.GetPageFromCache<NavigationViewModel>();

        navPage.PushAsync( (ContentPage) PageFactory.GetPageFromCache<IngredientViewModel>() );



        AddEntryCommand = new PageFactoryCommand(() =>

                                   navPage.PushModalPage()

                                   );
```

Btw. The little Donation this morning was from me

Von: Daniel Luberda [mailto:notifications@github.com] 
Gesendet: Mittwoch, 13. Januar 2016 14:26
An: daniel-luberda/DLToolkit.PageFactory DLToolkit.PageFactory@noreply.github.com
Cc: escamoteur thomas@burkharts.net
Betreff: Re: [DLToolkit.PageFactory] Navigation bar back button interception is not implemented (#6)

@escamoteur https://github.com/escamoteur  Not much, but it would handle your case. CurrentPage view model would be also a NavigationPage view model. And by looking at Navigation Page definition, it would be obvious that for this navigation page, current page wouldn't be changed at runtime.

**Comment by daniel-luberda** [comment]
@escamoteur Thanks! Didn't know about it, because didn't receive it yet. Paypal can process those kind of payments for about 24 hours. I appreciate it.

I was thinking about the problem and have a little idea how could we simplify this:

I could create PageFactory methods:

```
GetNavigationFromCache<TNavigationViewModel>
GetMessagableNavigationFromCache<TNavigationViewModel>
//etc
```

then modify extension methods, so calling PushPage(), PopPage  on  navigation page will use current navigation page navigation instead a global one. Calling the same methods on all other pages types will use global navigation. 

What do you think? Will it be more intuitive?

**Comment by escamoteur** [comment]
I’m not sure what would be the difference, because I’m still Pushing and Popping from a stack. 

This might be more useful if the NavigationPage is child of a TabbedPage or CarouselPage

Was my approach that I sent you the right one?

Von: Daniel Luberda [mailto:notifications@github.com] 
Gesendet: Mittwoch, 13. Januar 2016 15:53
An: daniel-luberda/DLToolkit.PageFactory DLToolkit.PageFactory@noreply.github.com
Cc: escamoteur thomas@burkharts.net
Betreff: Re: [DLToolkit.PageFactory] Navigation bar back button interception is not implemented (#6)

@escamoteur https://github.com/escamoteur  Thanks! Didn't know about it, because didn't receive it yet. Paypal can process those kind of payments for about 24 hours. 

I though about the problem and have a little idea how could we simplify this:

I could create PageFactory methods:

GetNavigationFromCache<TNavigationViewModel>
GetMessagableNavigationFromCache<TNavigationViewModel>
//etc

then modify extension methods, so calling PushPage(), PopPage on navigation page will use current navigation page navigation instead a global one. Calling the same methods on all other pages types will use global navigation. 

What do you think? Will it be more intuitive?

**Comment by daniel-luberda** [comment]
Could you elaborate if you're using just one navigation page? Or you have another navigation page which is deeper in hierarchy? 

The difference is that every NavigationPage has its own Navigation stack. Currently all PageFactory extension methods apply to global navigation. 

If you want to have a page which is a second navigation. You should push/pop pages on its internal navigation, not the global one. 

If you give me some more info, I'll try to make an example how to use it then :) Because currently I don't know which case is this.

**Comment by daniel-luberda** [comment]
## The first case (you only use one, global navigation):

``` C#
public IngredientsMainViewModel()
{
    AddEntryCommand = new PageFactoryCommand(() => {

        PageFactory.GetPageFromCache<IngredientViewModel>()
            .PushModalPage();

    });
}
```

## The second case (second navigation which is deeper in hierarchy than the global one):

### Currently possible:

``` C#
public IngredientsMainViewModel()
{   
    AddEntryCommand = new PageFactoryCommand(() => {
        var secondNav = (SecondNavigationPage)PageFactory.GetPageFromCache<SecondNavigationViewModel>();
        var pageToPush = PageFactory.GetPageFromCache<IngredientViewModel>();
        secondNav.PushModalAsync(pageToPush);

    });
}
```

### If I make suggested enhancements to PageFactory API, then it would be possible to do it more clearly:

``` C#
public IngredientsMainViewModel()
{
    AddEntryCommand = new PageFactoryCommand(() => {

        var pageToPush = PageFactory.GetPageFromCache<IngredientViewModel>();
        PageFactory.GetNavigationFromCache<SecondNavigationViewModel>()
            .PushModalPage(pageToPush);

    });
}
```

**Comment by escamoteur** [comment]
Could you elaborate if you're using just one navigation page? Or you have another navigation page which is deeper in hierarchy? 

I use a second navigation page deeper in the hierachy, but only to get a actionbar in a modal page. Not because I need there a new navigational stack.

The difference is that every NavigationPage has its own Navigation stack. Currently all PageFactory extension methods apply to global navigation. 

I understand that, but if I pop all Pages from a deeper Navigation page I will get back tot he main Navigation , don’t I? So in a way it’s just one big stack.

**Comment by escamoteur** [comment]
I don’t think the enhanced version is more intuitive as it introduces a new special Method for this case. Reading the Code from the current version describes very clear what happens.

The first case (you only use one, global navigation):

public IngredientsMainViewModel()
{
    AddEntryCommand = new PageFactoryCommand(() => {

```
    PageFactory.GetPageFromCache<IngredientViewModel>()
        .PushModalPage();

});
```

}

The second case (second navigation which is deeper in hierarchy than the global one):

Currently possible:

public IngredientsMainViewModel()
{  
    AddEntryCommand = new PageFactoryCommand(() => {
        var secondNav = (SecondNavigationViewModel)PageFactory.GetPageFromCache<SecondNavigationViewModel>();
        var pageToPush = PageFactory.GetPageFromCache<IngredientViewModel>();
        secondNav.PushModalAsync(pageToPush);

```
});
```

}

If I make suggested enhancements to PageFactory API, then it would be possible to do it more clearly:

public IngredientsMainViewModel()
{
    AddEntryCommand = new PageFactoryCommand(() => {

```
    var pageToPush = PageFactory.GetPageFromCache<IngredientViewModel>();
    PageFactory.GetNavigationFromCache<SecondNavigationViewModel>()
        .PushModalPage(pageToPush);

});
```

}

**Comment by escamoteur** [comment]
Getting back to your proposal

``` c#
public abstract class PFNavigationPage<TNavigationViewModel, TPageViewModel> : Xamarin.Forms.NavigationPage, IBasePage<TViewModel> where TViewModel : class, INotifyPropertyChanged
```

Does this mean, I inherit a new NavigationPage from this class
and when I call 

``` c#
PageFactory.GetPageFromCache<TPageViewModel>()
```

I will get the embedded ContentPage?

BTW. I don't need Nuget, I can now make a GIT-Pull

**Comment by daniel-luberda** [comment]
@escamoteur 

With my proposed changes:

``` C#
// this will get you a a navigation page or content page (we have to decide):
YourNavigationPage navigation = (YourNavigationPage)PageFactory.GetPageFromCache<YourNavigationViewModel>();
var page = PageFactory.GetPageFromCache<YourNavigationViewModel>();
// In my opinion it should return navigation page

// this will get you a a navigation page (without casting):
YourNavigationPage navigation = PageFactory.GetNavigationFromCache<YourNavigationViewModel>();

// this will get you a content page inside navigation page
YourNavigationPage navigation = PageFactory.GetNavigationFromCache<YourNavigationViewModel>();
var contentPage = navigation.CurrentPage; 
// or IBasePage type
var contentPage = navigation.GetPage();
// or IBasePage messagable type
var contentPage = navigation.GetMessagablePage();
```

> I don’t think the enhanced version is more intuitive as it introduces a new special Method for this case. Reading the Code from the current version describes very clear what happens.

The downside of current code is that bypassing PageFactory methods disables some features like navigation interception, etc. (`secondNav.PushModalAsync` is a pure Forms method). With proposed changed it would work fine. 

What do you think?

**Comment by daniel-luberda** [comment]
The other idea is force pages to always use their navigations: 
- If a page was pushed from global navigation it will always use global navigation
- If  a page is a deeper nested navigation type it will use itself to push pages
- If a page was pushed from deeper nested navigation, then it will use deeper nested navigation for navigating

Currently all PageFactory methods use global navigation.

**Comment by escamoteur** [comment]
> The downside of current code is that bypassing PageFactory methods disables some features like >navigation interception, etc. (secondNav.PushModalAsync is a pure Forms method). With proposed >changed it would work fine.

Why do you have to use PushModalAsync?

``` c#
        public IngredientsMainViewModel()
        {
            var navPage = (BREADy.Views.NavigationPage) PageFactory.GetPageFromCache<NavigationViewModel>();
            navPage.PushAsync( (ContentPage) PageFactory.GetPageFromCache<IngredientViewModel>() );

            AddEntryCommand = new PageFactoryCommand(() =>
                                       navPage.PushModalPage()
                                       );

        }
```

Here I only use PageFactory Methods

**Comment by escamoteur** [comment]
With your proposed changes I can reference the NavigationPage by it's viewmodel fine. 
But how does this help in the steps to create a new in a Navigationpage embedded Contentpage and Push both of them?

**Comment by daniel-luberda** [comment]
> Why do you have to use PushModalAsync?

@escamoteur Because, in current PageFactory version all navigation methods are called on global navigation page (the one which is specified in `XamarinPageFactory.Init` method). If you use another navigation which is nested you want to call navigation methods on it instead on global one.

**Comment by daniel-luberda** [comment]
@escamoteur With this: https://github.com/daniel-luberda/DLToolkit.PageFactory/issues/6#issuecomment-171352323 - the API won't change but it will use correct Navigation (global or nested if the page was pushed by nested navigation or is nested navigation itself).

**Comment by daniel-luberda** [comment]
> But how does this help in the steps to create a new in a Navigationpage embedded Contentpage and Push both of them?

When using this: 

``` C#
public abstract class PFNavigationPage<TNavigationViewModel, TPageViewModel> : Xamarin.Forms.NavigationPage, IBasePage<TViewModel> where TViewModel : class, INotifyPropertyChanged
```

You would just create your navigation page like this: 

``` C#
public class YourCustomNavigationPage : PFNavigationPage<NavigationViewModel, PageViewModel>
{   
}
```

this will use a different constructor for a NavigationPage when doing `PageFactory.GetNavigationFromCache<YourCustomNavigationPage>`. The second one: 
- NavigationPage() - Initializes a new NavigationPage object.
- NavigationPage(Page) - Creates a new NavigationPage element with root as its root element.

and you wouldn't need to push anything as PageViewModel Page will be a already a navigation root.

**Comment by escamoteur** [comment]
So it would automatically create the PageViewModel associated PFContentPage when it is created?

**Comment by daniel-luberda** [comment]
@escamoteur  Yes.

**Comment by daniel-luberda** [comment]
@escamoteur  I'm doing some changes now. I'll let you know when I commit it.

**Comment by escamoteur** [comment]
Hi Daniel,
any progress on your side?
Best
Thomas

**Comment by daniel-luberda** [comment]
@escamoteur Yes, quite a large commit, but have to finish some things first. Got some private urgent matter, and currently I don't have an access to my dev machine. I'll let you know when i do a commit. It should be really soon, maybe even today.

**Comment by daniel-luberda** [comment]
@escamoteur I though I'm responding to another issue (sub navigation pages). 

As to back navigation handling I have some ideas how to achieve that. The only requirement would be to use PageFactory methods for navigation. But it involves some timers and I would prefer to avoid it.

**Comment by escamoteur** [comment]
I have not yet asked the Xamarin support about the Issue. I will do.

Von: Daniel Luberda [mailto:notifications@github.com] 
Gesendet: Dienstag, 19. Januar 2016 17:54
An: daniel-luberda/DLToolkit.PageFactory
Cc: escamoteur
Betreff: Re: [DLToolkit.PageFactory] Navigation bar back button interception is not implemented (#6)

@escamoteur https://github.com/escamoteur  I though I'm responding to another issue (sub navigation pages). 

As to back navigation handling I have some ideas how to achieve that. The only requirement would be to use PageFactory methods for navigation. But it involves some timers and I would prefer to avoid it.

**Comment by escamoteur** [comment]
Hi Daniel,

did you check on this on XF 2.0? I just ask, because before I ask Xamarin support want be sure that this is still the case.

Best

Thomas Burkhart

Von: Daniel Luberda [mailto:notifications@github.com] 
Gesendet: Dienstag, 19. Januar 2016 17:54
An: daniel-luberda/DLToolkit.PageFactory
Cc: escamoteur
Betreff: Re: [DLToolkit.PageFactory] Navigation bar back button interception is not implemented (#6)

@escamoteur https://github.com/escamoteur  I though I'm responding to another issue (sub navigation pages). 

As to back navigation handling I have some ideas how to achieve that. The only requirement would be to use PageFactory methods for navigation. But it involves some timers and I would prefer to avoid it.

**Comment by daniel-luberda** [comment]
@escamoteur Yes, nothing has changed in 2.0 when it comes to navigation back interception.

**Comment by escamoteur** [comment]
Hi Daniel,

did you check for 2.0 if also the strange behaviour oft he stack still exists?

Von: Daniel Luberda [mailto:notifications@github.com] 
Gesendet: Mittwoch, 20. Januar 2016 20:53
An: daniel-luberda/DLToolkit.PageFactory
Cc: escamoteur
Betreff: Re: [DLToolkit.PageFactory] Navigation bar back button interception is not implemented (#6)

@escamoteur https://github.com/escamoteur  Yes, nothing has changed in 2.0 when it comes to navigation back interception.

**Comment by escamoteur** [comment]
Hi Daniel,

do you have any sample project that shows the strange NavigationStack behavior? I would like to send the issue to Xamarin support, but the always ask for some sample project that shows the effect.

**Comment by daniel-luberda** [comment]
The simplest project with two pages `FirstPage`, `SecondPage`:
- In `FirstPage` create a button which pushes `SecondPage`
- In `FirstPage` `OnAppearing` method just add a code that logs `GetType()` of NavigationPage.NavigationStack items. 

Then, to reproduce run it (every platform has a little different behaviour):
-  Open  `SecondPage` navigate back with navigation bar. Result: `SecondPage` is still in NavigationStack and disappears only after another page is pushed. UNEXPECTED BEHAVIOUR
-  Open  `SecondPage` navigate back with hardware back button. Result: Navigation stack doesn't contain `SecondPage` type.

If you make a repro, feel free to mail it to me. I'll test it too with the newest XF.

**Comment by escamoteur** [comment]
:-) Thought you could make one :-)

**Comment by daniel-luberda** [comment]
Sure, I can do that, but first I need to catch up with some other things.

**Comment by escamoteur** [comment]
It's not urgent. I just thought it may help if I sent it to Xamarin support.

**Comment by cbfrank** [comment]
I found that Page.Parent is the NavigationPage, and it has the Popped event, which will be called when the page is popped when user click back button, maybe use this as a hook?
